source/ada/lsp-ada_handlers.adb

```diff
@@ -431,14 +434,206 @@ package body LSP.Ada_Handlers is
       Value : LSP.Messages.TextDocumentPositionParams)
       return LSP.Messages.Hover_Response
    is
-      pragma Unreferenced (Self, Value);
-      Response : LSP.Messages.Hover_Response (Is_Error => True);
+      use Libadalang.Analysis;
+      use Libadalang.Common;
+
+      Response           : LSP.Messages.Hover_Response (Is_Error => False);
+      Document           : constant LSP.Ada_Documents.Document_Access :=
+                             Self.Context.Get_Document
+                               (Value.textDocument.uri);
+      Name_Node          : constant Name :=
+                             LSP.Lal_Utils.Get_Node_As_Name
+                               (Document.Get_Node_At (Value.position));
+      Defining_Name_Node : Defining_Name;
+      Decl               : Basic_Decl;
+      Subp_Spec_Node     : Base_Subp_Spec;
+      Hover_Text         : LSP_String;
+
+      procedure Create_Hover_Text_For_Basic_Decl;
+      --  Create the hover text for for basic declarations
+
+      procedure Create_Hover_Text_For_Subp_Spec;
+      --  Create the hover text for for subprogram declarations
+
+      --------------------------------------
+      -- Create_Hover_Text_For_Basic_Decl --
+      --------------------------------------
+
+      procedure Create_Hover_Text_For_Basic_Decl is
+      begin
+         --  Return an empty hover text for package declarations
+         if Decl.Kind in Ada_Base_Package_Decl then
+            return;
+         end if;
+
+         declare
+            Text        : constant String := Langkit_Support.Text.To_UTF8
+              (Decl.Text);
+            Lines       : GNAT.Strings.String_List_Access := Split
+              (Text,
+               On               => ASCII.LF,
+               Omit_Empty_Lines => True);
+            Idx         : Integer;
+         begin
+            --  If it's a single-line declaration, replace all the series of
+            --  whitespaces by only one blankspace.
+            --  If it's a multi-line declaration, remove only the unneeded
+            --  indentation whitespaces.
+
```

Comment:
for safety, should we add a case if Text = "" ? (it would only serve as defensive code, I don't think it can happen in real life)

Reply:
Ok will do